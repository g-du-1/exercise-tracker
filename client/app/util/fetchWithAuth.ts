export const fetchWithAuth = async (url: string, options: RequestInit = {}) => {
  const baseUrl = process.env.NEXT_PUBLIC_API_PREFIX || "/api/v1";

  options.credentials = "include";

  options.headers = {
    ...(options.headers as Record<string, string>),
    "Content-Type": "application/json",
    Accept: "application/json",
  };

  const resp = await fetch(`${baseUrl}${url}`, options);

  if (resp?.status === 401) {
    throw new Error("401: Unauthorised");
  }

  if (resp?.status !== 200) {
    throw new Error("Something went wrong.");
  }

  return resp;
};
