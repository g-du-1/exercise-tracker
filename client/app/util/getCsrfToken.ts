export const getCsrfToken = async (baseUrl: string) => {
  const csrfEndpoint =
    process.env.NEXT_PUBLIC_API_CSRF_ENDPOINT || "/csrf-token";

  const response = await fetch(`${baseUrl}${csrfEndpoint}`, {
    credentials: "include",
  });

  const data = await response?.json();

  return data?.token;
};
