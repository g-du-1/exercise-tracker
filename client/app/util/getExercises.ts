export const getExercises = async () => {
  // TODO Make it an env var
  const res = await fetch("http://localhost:8080/api/v1/exercises");

  return res.json();
};
