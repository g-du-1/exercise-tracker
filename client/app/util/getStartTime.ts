export const getStartTime = (date = new Date()) => {
  return date.toLocaleTimeString([], {
    hour: "2-digit",
    minute: "2-digit",
    hour12: false,
  });
};
