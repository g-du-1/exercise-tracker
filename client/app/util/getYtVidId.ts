export const getYtVidId = (ytLink: string) => {
  const urlParams = new URLSearchParams(ytLink.split("?")[1]);
  return urlParams.get("v");
};
