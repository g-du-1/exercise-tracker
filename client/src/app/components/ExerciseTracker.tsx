"use client";

import { Exercise } from "../types";
import IconButton from "@mui/material/IconButton";
import AddIcon from "@mui/icons-material/Add";
import Box from "@mui/material/Box";
import * as React from "react";
import Typography from "@mui/material/Typography";
import Modal from "@mui/material/Modal";

const style = {
  position: "absolute" as "absolute",
  top: "50%",
  left: "50%",
  transform: "translate(-50%, -50%)",
  width: 400,
  bgcolor: "background.paper",
  p: 4,
};

export const ExerciseTracker = ({ exercises }: { exercises: Exercise[] }) => {
  const [selectedExercise, setSelectedExercise] = React.useState("");
  const [modalOpen, setModalOpen] = React.useState(false);

  const handleAddClick = (selectedExercise: string) => {
    setSelectedExercise(selectedExercise);
    setModalOpen(true);
  };

  const handleModalClose = () => setModalOpen(false);

  return (
    <>
      <Box>
        {exercises.map((exercise) => (
          <Box key={exercise.id} sx={{ display: "flex", alignItems: "center" }}>
            <Box>{exercise.name}</Box>

            <IconButton
              aria-label="add"
              onClick={() => handleAddClick(exercise.name)}
            >
              <AddIcon />
            </IconButton>
          </Box>
        ))}
      </Box>

      <Modal open={modalOpen} onClose={handleModalClose}>
        <Box sx={style}>
          <Typography variant="h6" component="h2">
            Add {selectedExercise} Reps
          </Typography>

          {/* <Box sx={{ mt: 2 }}>Description placeholder</Box> */}
        </Box>
      </Modal>
    </>
  );
};
