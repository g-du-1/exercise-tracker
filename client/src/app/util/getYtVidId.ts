export const getYtVidId = (ytLink: string) => {
  const urlParams = new URLSearchParams(ytLink.split("?")[1]);
  const videoId = urlParams.get("v");

  return videoId;
};
