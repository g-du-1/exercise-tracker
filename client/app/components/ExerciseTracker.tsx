"use client";

import { Exercise } from "../types";
import IconButton from "@mui/material/IconButton";
import AddIcon from "@mui/icons-material/Add";
import Box from "@mui/material/Box";
import * as React from "react";
import { getYtVidId } from "../util/getYtVidId";
import InfoIcon from "@mui/icons-material/Info";
import Divider from "@mui/material/Divider";
import TextField from "@mui/material/TextField";
import Button from "@mui/material/Button";
import Dialog from "@mui/material/Dialog";
import DialogActions from "@mui/material/DialogActions";
import DialogContent from "@mui/material/DialogContent";
import DialogTitle from "@mui/material/DialogTitle";
import ClearIcon from "@mui/icons-material/Clear";
import { useRef, useState } from "react";
import PlayArrowIcon from "@mui/icons-material/PlayArrow";
import RestartAltIcon from "@mui/icons-material/RestartAlt";
import DoneIcon from "@mui/icons-material/Done";

type SavedReps = {
  name: string;
  reps: number[];
};

export const ExerciseTracker = ({ exercises }: { exercises: Exercise[] }) => {
  const [fieldValue, setFieldValue] = useState("");
  const [savedReps, setSavedReps] = useState<{
    [key: string]: SavedReps;
  }>({});
  const [showMoreInfo, setShowMoreInfo] = useState(false);
  const [selectedExercise, setSelectedExercise] = useState<Exercise | null>(
    null
  );
  const [modalOpen, setModalOpen] = useState(false);

  const [swRunning, setSwRunning] = useState(false);
  const [swElapsedTime, setSwElapsedTime] = useState(0);
  const swIntervalRef = useRef<NodeJS.Timeout | null>(null);

  const startStopwatch = () => {
    setSwRunning(true);

    swIntervalRef.current = setInterval(() => {
      setSwElapsedTime((prevElapsedTime) => prevElapsedTime + 1);
    }, 1000);
  };

  const resetStopwatch = () => {
    if (swIntervalRef.current) clearInterval(swIntervalRef.current);

    setSwElapsedTime(0);
    setSwRunning(false);
  };

  const formatSwTime = (timeInSeconds: number) => {
    const hours = Math.floor(timeInSeconds / 3600);
    const minutes = Math.floor((timeInSeconds % 3600) / 60);
    const seconds = timeInSeconds % 60;

    return `${hours.toString().padStart(2, "0")}:${minutes
      .toString()
      .padStart(2, "0")}:${seconds.toString().padStart(2, "0")}`;
  };

  const handleAddClick = (selectedExercise: Exercise) => {
    setSelectedExercise(selectedExercise);
    setModalOpen(true);
  };

  const handleModalClose = () => {
    const fieldValNum = parseInt(fieldValue);

    if (selectedExercise && fieldValNum) {
      const newReps = { ...savedReps };

      const exercise = newReps[selectedExercise.id];

      if (exercise) {
        exercise.reps.push(parseInt(fieldValue));
      } else {
        newReps[selectedExercise.id] = {
          name: selectedExercise.name,
          reps: [fieldValNum],
        };
      }

      setSavedReps(newReps);

      if (selectedExercise.category !== "warmup") {
        resetStopwatch();
        startStopwatch();
      }
    }

    setModalOpen(false);
    setFieldValue("");
  };

  return (
    <>
      <Box
        sx={{
          display: "flex",
          justifyContent: "space-between",
          position: "fixed",
          top: 0,
          zIndex: 1,
          background: "white",
          width: "100%",
          opacity: 0.9,
          left: "50%",
          transform: "translateX(-50%)",
          borderBottom: "1.5px solid #e8e8e8",
        }}
      >
        <Box
          sx={{ display: "flex", flexDirection: "row", alignItems: "center" }}
        >
          <IconButton
            color="success"
            size="large"
            aria-label="Start Stopwatch"
            disabled={swRunning}
            onClick={startStopwatch}
          >
            <PlayArrowIcon />
          </IconButton>

          <IconButton
            size="large"
            color="error"
            aria-label="Reset Stopwatch"
            disabled={!swRunning}
            onClick={resetStopwatch}
          >
            <RestartAltIcon />
          </IconButton>

          <Box sx={{ ml: 1 }}>{formatSwTime(swElapsedTime)}</Box>
        </Box>

        <IconButton
          size="large"
          aria-label="Show More Info"
          onClick={() => setShowMoreInfo(!showMoreInfo)}
        >
          <InfoIcon />
        </IconButton>
      </Box>

      <Box>
        {exercises.map((exercise, idx) => (
          <Box
            key={exercise.id}
            sx={{
              display: "flex",
              alignItems: "center",
              flexDirection: "column",
            }}
          >
            <Box
              sx={{
                display: "flex",
                flexDirection: "column",
                width: "100%",
              }}
            >
              <Box
                sx={{
                  display: "flex",
                  alignItems: "center",
                  width: "100%",
                  justifyContent: "space-between",
                }}
              >
                <Box sx={{ display: "flex", alignItems: "center" }}>
                  <Box>{exercise.name}</Box>

                  {savedReps?.[exercise.id]?.reps.length >=
                    exercise.targetSets && (
                    <IconButton color="success" size="small" aria-label="Done">
                      <DoneIcon />
                    </IconButton>
                  )}
                </Box>

                <Box>
                  {savedReps?.[exercise.id]?.reps?.length > 0 && (
                    <IconButton
                      color="error"
                      size="large"
                      aria-label="Delete Reps"
                      onClick={() => {
                        const newReps = { ...savedReps };

                        const existingExercise = newReps[exercise.id];

                        if (existingExercise) {
                          existingExercise.reps = [];
                        }

                        setSavedReps(newReps);
                      }}
                    >
                      <ClearIcon />
                    </IconButton>
                  )}

                  <IconButton
                    size="large"
                    aria-label="add"
                    onClick={() => handleAddClick(exercise)}
                  >
                    <AddIcon />
                  </IconButton>
                </Box>
              </Box>

              <Box sx={{ display: "flex", flexDirection: "row" }}>
                {savedReps?.[exercise.id]?.reps.map((rep: number, idx) => (
                  <Box
                    key={`${exercise.id}-${idx}`}
                    sx={{ display: "flex", flexDirection: "row", mr: 1, mb: 2 }}
                  >
                    <Box sx={{ mr: 0.75, color: "#b9b9b9" }}>
                      Set {idx + 1}:
                    </Box>

                    <Box sx={{ fontWeight: "500" }}>{rep}</Box>
                  </Box>
                ))}
              </Box>
            </Box>

            {exercise.thumbLink &&
              (exercise.thumbLink.includes("youtube") ? (
                <Box
                  className="video-responsive"
                  sx={{ mb: 2 }}
                  style={!showMoreInfo ? { display: "none" } : {}}
                >
                  <iframe
                    src={`https://www.youtube.com/embed/${getYtVidId(
                      exercise.thumbLink
                    )}`}
                    allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
                    allowFullScreen
                    title={`${exercise.name} Video`}
                  />
                </Box>
              ) : exercise.thumbLink.endsWith("jpg") ? (
                <img
                  src={exercise.thumbLink}
                  alt={`${exercise.name} Image`}
                  style={
                    !showMoreInfo
                      ? { display: "none" }
                      : { width: "100%", height: "auto" }
                  }
                />
              ) : null)}

            {showMoreInfo && exercise.comments && (
              <Box
                dangerouslySetInnerHTML={{ __html: exercise.comments }}
                sx={{ width: "100%" }}
              />
            )}

            {exercises[idx + 1] &&
              exercise.category !== exercises[idx + 1].category && (
                <Divider flexItem data-testid="divider" sx={{ my: 1 }} />
              )}
          </Box>
        ))}
      </Box>

      <Dialog open={modalOpen} onClose={handleModalClose} disableRestoreFocus>
        <DialogTitle>Add {selectedExercise?.name} Reps</DialogTitle>

        <DialogContent>
          <TextField
            autoFocus
            margin="dense"
            label="Reps"
            type="number"
            fullWidth
            variant="standard"
            value={fieldValue}
            onChange={(e) => setFieldValue(e.target.value)}
          />
        </DialogContent>

        <DialogActions>
          <Button onClick={() => setModalOpen(false)}>Cancel</Button>
        </DialogActions>
      </Dialog>
    </>
  );
};
