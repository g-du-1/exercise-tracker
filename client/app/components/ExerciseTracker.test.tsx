/**
 * @jest-environment jsdom
 */
import { render, fireEvent, screen, act } from "@testing-library/react";
import { ExerciseTracker } from "./ExerciseTracker";
import { getStartTime } from "../util/getStartTime";
import { mockExercises } from "./fixtures/mockExercises";

jest.useFakeTimers();

const DELETE_REPS = "Delete Reps";

const renderExerciseTracker = () =>
  render(<ExerciseTracker exercises={mockExercises} />);

const getModalOpenBtns = () => screen.getAllByLabelText("add");

const clickOpenModalBtn = (idx: number) => {
  const openModalBtns = getModalOpenBtns();

  fireEvent.click(openModalBtns[idx]);
};

const getRepsInput = () => {
  const input = screen.getByLabelText("Reps");
  expect(input).toHaveFocus();

  return input;
};

const submitReps = (val: string) => {
  const input = getRepsInput();
  fireEvent.change(input, { target: { value: val } });
  fireEvent.submit(input);
};

const getDeleteRepsBtn = () => screen.getByLabelText(DELETE_REPS);

const clickDeleteReps = () => fireEvent.click(getDeleteRepsBtn());

describe("ExerciseTracker", () => {
  it("matches snapshot", () => {
    const result = renderExerciseTracker();

    expect(result.container).toMatchSnapshot();
  });

  it("displays modal open buttons next to each exercise", () => {
    renderExerciseTracker();

    expect(getModalOpenBtns()).toHaveLength(3);
  });

  it("open modal button opens a modal on click for the exercise", () => {
    renderExerciseTracker();

    fireEvent.click(getModalOpenBtns()[0]);

    expect(screen.getByText("Add GMB Wrist Prep Reps")).toBeDefined();
  });

  it("does not display more info by default", () => {
    renderExerciseTracker();

    expect(screen.queryByTitle("GMB Wrist Prep Video")).not.toBeVisible();
    expect(screen.queryByTitle("Arch Hangs Video")).not.toBeVisible();
    expect(
      screen.queryByAltText("Parallel Bar Support Hold Image")
    ).not.toBeVisible();
  });

  it("displays videos and images when more info is clicked", () => {
    renderExerciseTracker();

    fireEvent.click(screen.getByLabelText("Show More Info"));

    expect(screen.getByTitle("GMB Wrist Prep Video")).toBeVisible();
    expect(screen.getByTitle("Arch Hangs Video")).toBeVisible();
    expect(
      screen.getByAltText("Parallel Bar Support Hold Image")
    ).toBeVisible();
  });

  it("renders category separators when category changes", () => {
    renderExerciseTracker();

    expect(screen.getAllByTestId("divider")).toHaveLength(2);
  });

  it("saves and resets sets of reps for an exercise", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    submitReps("7");

    clickOpenModalBtn(1);

    submitReps("6");

    expect(screen.getByText("7")).toBeInTheDocument();
    expect(screen.getByText("6")).toBeInTheDocument();

    clickOpenModalBtn(1);

    clickDeleteReps();

    expect(screen.queryByText("7")).not.toBeInTheDocument();
    expect(screen.queryByText("6")).not.toBeInTheDocument();
  });

  it("saves reps on modal close", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    const repsInput = getRepsInput();

    fireEvent.change(repsInput, { target: { value: "8" } });

    fireEvent.keyDown(repsInput, {
      key: "Escape",
      code: "Escape",
    });

    expect(screen.getByText("8")).toBeInTheDocument();
  });

  it("only renders the delete reps icon if there are reps saved and closes the modal on save", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    expect(screen.queryByLabelText(DELETE_REPS)).not.toBeInTheDocument();

    submitReps("6");

    clickOpenModalBtn(1);

    expect(screen.getByLabelText(DELETE_REPS)).toBeInTheDocument();

    clickDeleteReps();

    expect(screen.queryByText("Add Arch Hangs Reps")).not.toBeVisible();
  });

  it("starts the stopwatch when saving reps", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    submitReps("6");

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(screen.getByText("00:00:05")).toBeInTheDocument();
  });

  it("does not start the stopwatch when saving a warmup", () => {
    renderExerciseTracker();

    clickOpenModalBtn(0);

    submitReps("6");

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(screen.getByText("00:00:00")).toBeInTheDocument();
  });

  it("starts the stopwatch", () => {
    renderExerciseTracker();

    const startBtn = screen.getByLabelText("Start Stopwatch");

    fireEvent.click(startBtn);

    expect(startBtn).toBeDisabled();

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    expect(screen.getByText("00:00:05")).toBeInTheDocument();
  });

  it("resets the stopwatch", () => {
    renderExerciseTracker();

    const resetBtn = screen.getByLabelText("Reset Stopwatch");

    expect(resetBtn).toBeDisabled();

    const startBtn = screen.getByLabelText("Start Stopwatch");

    fireEvent.click(startBtn);

    expect(startBtn).toBeDisabled();

    act(() => {
      jest.advanceTimersByTime(8000);
    });

    fireEvent.click(resetBtn);

    expect(screen.getByText("00:00:00")).toBeInTheDocument();
  });

  it("displays a done elem when the saved sets reach the target sets", () => {
    renderExerciseTracker();

    expect(screen.queryByLabelText("Done")).not.toBeInTheDocument();

    clickOpenModalBtn(0);

    submitReps("6");

    expect(screen.getByLabelText("Done")).toBeInTheDocument();
  });

  it("displays warning after the target rest has passed", () => {
    renderExerciseTracker();

    expect(screen.queryByAltText("Rest Time Passed")).not.toBeInTheDocument();

    clickOpenModalBtn(1);

    submitReps("6");

    act(() => {
      jest.advanceTimersByTime(90000);
    });

    expect(screen.getByLabelText("Rest Time Passed")).toBeInTheDocument();
  });

  it("resets the timer when reps are deleted", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    submitReps("6");

    act(() => {
      jest.advanceTimersByTime(5000);
    });

    clickOpenModalBtn(1);

    clickDeleteReps();

    expect(screen.getByText("00:00:00")).toBeInTheDocument();
  });

  it("displays target sets", () => {
    renderExerciseTracker();

    expect(screen.getByText("1x10")).toBeInTheDocument();
    expect(screen.getByText("3x5-8")).toBeInTheDocument();
    expect(screen.getByText("3x30s")).toBeInTheDocument();
  });

  it("adds label for in range reps", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    submitReps("6");

    expect(
      screen.getByLabelText("Arch Hangs 6 Reps In Range")
    ).toBeInTheDocument();
  });

  it("adds label for out of range reps - lower", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    submitReps("3");

    expect(
      screen.getByLabelText("Arch Hangs 3 Reps Lower Than Range")
    ).toBeInTheDocument();
  });

  it("adds label for out of range reps - higher", () => {
    renderExerciseTracker();

    clickOpenModalBtn(1);

    submitReps("35");

    expect(
      screen.getByLabelText("Arch Hangs 35 Reps Higher Than Range")
    ).toBeInTheDocument();
  });

  it("adds label for cases without max target step", () => {
    renderExerciseTracker();

    clickOpenModalBtn(0);

    submitReps("10");

    expect(
      screen.getByLabelText("GMB Wrist Prep 10 Reps In Range")
    ).toBeInTheDocument();
  });

  it("hides completed exercises if hiding is turned on", () => {
    renderExerciseTracker();

    expect(screen.getByText("GMB Wrist Prep")).toBeVisible();
    expect(screen.getByText("Arch Hangs")).toBeVisible();
    expect(screen.getByText("Parallel Bar Support Hold")).toBeVisible();

    clickOpenModalBtn(0);

    submitReps("8");

    const checkbox = screen.getByLabelText("Show Completed Exercises");
    expect(checkbox).toBeChecked();

    fireEvent.click(checkbox);
    expect(checkbox).not.toBeChecked();

    expect(screen.getByText("GMB Wrist Prep")).not.toBeVisible();
    expect(screen.getByText("Arch Hangs")).toBeVisible();
    expect(screen.getByText("Parallel Bar Support Hold")).toBeVisible();
  });

  it("saves and renders start time", () => {
    renderExerciseTracker();

    clickOpenModalBtn(0);

    submitReps("8");

    const startTime = getStartTime();

    expect(screen.getByText(`Started: ${startTime}`)).toBeInTheDocument();

    act(() => {
      jest.advanceTimersByTime(500000);
    });

    clickOpenModalBtn(1);

    submitReps("6");

    expect(screen.getByText(`Started: ${startTime}`)).toBeInTheDocument();
  });
});
