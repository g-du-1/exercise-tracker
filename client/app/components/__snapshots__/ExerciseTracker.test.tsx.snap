// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExerciseTracker matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-1bf2odj"
  >
    <div
      class="MuiBox-root css-u4p24i"
    >
      <button
        aria-label="Start Stopwatch"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorSuccess MuiIconButton-sizeLarge css-93mzl9-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="PlayArrowIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M8 5v14l11-7z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
      <button
        aria-label="Reset Stopwatch"
        class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorError MuiIconButton-sizeLarge css-fz35w-MuiButtonBase-root-MuiIconButton-root"
        disabled=""
        tabindex="-1"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="RestartAltIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 5V2L8 6l4 4V7c3.31 0 6 2.69 6 6 0 2.97-2.17 5.43-5 5.91v2.02c3.95-.49 7-3.85 7-7.93 0-4.42-3.58-8-8-8m-6 8c0-1.65.67-3.15 1.76-4.24L6.34 7.34C4.9 8.79 4 10.79 4 13c0 4.08 3.05 7.44 7 7.93v-2.02c-2.83-.48-5-2.94-5-5.91"
          />
        </svg>
      </button>
      <div
        class="MuiBox-root css-1isemmb"
      >
        00:00:00
      </div>
    </div>
    <div
      class="MuiBox-root css-0"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium Mui-checked MuiCheckbox-root MuiCheckbox-colorDefault MuiCheckbox-sizeMedium css-dk24d0-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="Show Completed Exercises"
          checked=""
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="CheckBoxIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M19 3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.11 0 2-.9 2-2V5c0-1.1-.89-2-2-2zm-9 14l-5-5 1.41-1.41L10 14.17l7.59-7.59L19 8l-9 9z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <button
        aria-label="Show More Info"
        class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
        tabindex="0"
        type="button"
      >
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="InfoIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m1 15h-2v-6h2zm0-8h-2V7h2z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xaep3o-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiBox-root css-1nsk1m1"
      >
        <div
          class="video-responsive MuiBox-root css-0"
          style="display: none;"
        >
          <iframe
            allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
            allowfullscreen=""
            src="https://www.youtube.com/embed/mSZWSQSSEjE"
            title="GMB Wrist Prep Video"
          />
        </div>
      </div>
      <div
        class="MuiCardContent-root css-1v7oru7-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h2 css-1f5et6c-MuiTypography-root"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <div
              class="MuiBox-root css-70qvj9"
            >
              <div
                class="MuiBox-root css-12yyoxx"
              >
                1x10
              </div>
              <div
                class="MuiBox-root css-bg3st0"
              >
                GMB Wrist Prep
              </div>
            </div>
          </div>
          <button
            aria-label="add"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="OpenInNewIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-1n9ud6n"
        />
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-1y1k9n9-MuiDivider-root"
      data-testid="divider"
    />
  </div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xaep3o-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiBox-root css-1nsk1m1"
      >
        <div
          class="video-responsive MuiBox-root css-0"
          style="display: none;"
        >
          <iframe
            allow="accelerometer; autoplay; clipboard-write; encrypted-media; gyroscope; picture-in-picture"
            allowfullscreen=""
            src="https://www.youtube.com/embed/C995b3KLXS4"
            title="Arch Hangs Video"
          />
        </div>
      </div>
      <div
        class="MuiCardContent-root css-1v7oru7-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h2 css-1f5et6c-MuiTypography-root"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <div
              class="MuiBox-root css-70qvj9"
            >
              <div
                class="MuiBox-root css-12yyoxx"
              >
                3x5-8
              </div>
              <div
                class="MuiBox-root css-bg3st0"
              >
                Arch Hangs
              </div>
            </div>
          </div>
          <button
            aria-label="add"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="OpenInNewIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-1n9ud6n"
        />
      </div>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth MuiDivider-flexItem css-1y1k9n9-MuiDivider-root"
      data-testid="divider"
    />
  </div>
  <div
    class="MuiBox-root css-0"
  >
    <div
      class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-xaep3o-MuiPaper-root-MuiCard-root"
    >
      <div
        class="MuiBox-root css-1nsk1m1"
      >
        <img
          alt="Parallel Bar Support Hold Image"
          src="https://antranik.org/wp-content/uploads/2014/01/antranik-holding-support-hold-on-parallel-bars.jpg"
          style="display: none;"
        />
      </div>
      <div
        class="MuiCardContent-root css-1v7oru7-MuiCardContent-root"
      >
        <div
          class="MuiTypography-root MuiTypography-h2 css-1f5et6c-MuiTypography-root"
        >
          <div
            class="MuiBox-root css-70qvj9"
          >
            <div
              class="MuiBox-root css-70qvj9"
            >
              <div
                class="MuiBox-root css-12yyoxx"
              >
                3x30s
              </div>
              <div
                class="MuiBox-root css-bg3st0"
              >
                Parallel Bar Support Hold
              </div>
            </div>
          </div>
          <button
            aria-label="add"
            class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeLarge css-mf1cb5-MuiButtonBase-root-MuiIconButton-root"
            tabindex="0"
            type="button"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="OpenInNewIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M19 19H5V5h7V3H5c-1.11 0-2 .9-2 2v14c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2v-7h-2zM14 3v2h3.59l-9.83 9.83 1.41 1.41L19 6.41V10h2V3z"
              />
            </svg>
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
        <div
          class="MuiBox-root css-1n9ud6n"
        />
      </div>
    </div>
  </div>
</div>
`;
